projects/RabbitMQ.Client/client/impl/ContentHeaderPropertyReader.cs

```diff
@@ -78,82 +80,81 @@ public bool ReadBit()
-            m_flagWord = NetworkOrderDeserializer.ReadUInt16(_memory.Slice(_memoryOffset).Span);
-            _memoryOffset += 2;
-            m_bitCount = 0;
+            _bits = NetworkOrderDeserializer.ReadUInt16(Span);
+            _offset += 2;
+            _bitMask = StartBitMask;
         }
 
         public uint ReadLong()
         {
-            uint result = NetworkOrderDeserializer.ReadUInt32(_memory.Slice(_memoryOffset).Span);
-            _memoryOffset += 4;
+            uint result = NetworkOrderDeserializer.ReadUInt32(Span);
+            _offset += 4;
             return result;
         }
 
         public ulong ReadLonglong()
         {
-            ulong result = NetworkOrderDeserializer.ReadUInt64(_memory.Slice(_memoryOffset).Span);
-            _memoryOffset += 8;
+            ulong result = NetworkOrderDeserializer.ReadUInt64(Span);
+            _offset += 8;
             return result;
         }
 
         public byte[] ReadLongstr()
         {
-            byte[] result = WireFormatting.ReadLongstr(_memory.Slice(_memoryOffset));
-            _memoryOffset += 4 + result.Length;
+            byte[] result = WireFormatting.ReadLongstr(Span);
+            _offset += 4 + result.Length;
             return result;
         }
 
         public byte ReadOctet()
         {
-            return _memory.Span[_memoryOffset++];
+            return _span[_offset++];
         }
 
         public bool ReadPresence()
         {
-            if (m_bitCount == 15)
+            if (_bitMask == EndBitMask)
             {
-                ReadFlagWord();
+                ReadBits();
             }
 
-            int bit = 15 - m_bitCount;
-            bool result = (m_flagWord & (1 << bit)) != 0;
-            m_bitCount++;
+            bool result = (_bits & _bitMask) != 0;
+            _bitMask >>= 1;
             return result;
         }
 
         public ushort ReadShort()
         {
-            ushort result = NetworkOrderDeserializer.ReadUInt16(_memory.Slice(_memoryOffset).Span);
-            _memoryOffset += 2;
+            ushort result = NetworkOrderDeserializer.ReadUInt16(Span);
+            _offset += 2;
             return result;
         }
 
         public string ReadShortstr()
         {
-            string result = WireFormatting.ReadShortstr(_memory.Slice(_memoryOffset), out int bytesRead);
-            _memoryOffset += bytesRead;
+            string result = WireFormatting.ReadShortstr(Span, out int bytesRead);
+            _offset += bytesRead;
             return result;
         }
 
         /// <returns>A type of <seealso cref="System.Collections.Generic.IDictionary{TKey,TValue}"/>.</returns>
         public Dictionary<string, object> ReadTable()
         {
-            Dictionary<string, object> result = WireFormatting.ReadTable(_memory.Slice(_memoryOffset), out int bytesRead);
-            _memoryOffset += bytesRead;
+            Dictionary<string, object> result = WireFormatting.ReadTable(Span, out int bytesRead);
```

Comment:
Would we want to go down the path of actually pooling header dictionaries and then return it when the command is disposed as we do for the body buffer?

Reply:
Possible, but risky and a "breaking change" as the consumer would be prohibited from taking any reference to these passed dictionaries. (Maybe you can open an Issue and we can discuss there about pro / cons?

Reply:
Yeah true. Maybe it is not worth the risk though. I'm guessing a read-only dictionary would also not really help because someone might already somewhere abusing the writable nature of the type returned. So I'm on the fence if I even should raise an issue. @michaelklishin @lukebakken thoughts?

Reply:
I've thought about this pooling. Also, caching that default BasicProperties and only allocating for the fields that change from the defaults to minimize allocations taking place there. Requires some bookeeping around when the default properties change though but should be doable and yield a perf benefit.

Reply:
Pooling header dictionaries feels like a 7.0 thing. I think we'd want to prove that the benefits outweigh the (probably) more complicated code.

Reply:
Let's handle pooling in a separate PR (and yes, sounds like a 7.0 change which may or may not be worth the complexity).